src/Stripe.net/Infrastructure/Util.cs

```diff
@@ -0,0 +1,109 @@
+            { "order", typeof(Order) },
+            { "order_item", typeof(OrderItem) },
+            { "order_return", typeof(OrderReturn) },
+            { "payment_intent", typeof(PaymentIntent) },
+            { "payout", typeof(Payout) },
+            { "person", typeof(Person) },
+            { "plan", typeof(Plan) },
+            { "product", typeof(Product) },
+            { "recipient", typeof(Recipient) },
+            { "refund", typeof(Refund) },
+            { "review", typeof(Review) },
+            { "reporting.report_run", typeof(Reporting.ReportRun) },
+            { "reporting.report_type", typeof(Reporting.ReportType) },
+            { "scheduled_query_run", typeof(Sigma.ScheduledQueryRun) },
+            { "sku", typeof(Sku) },
+            { "source", typeof(Source) },
+            { "source_mandate_notification", typeof(SourceMandateNotification) },
+            { "source_transaction", typeof(SourceTransaction) },
+            { "subscription", typeof(Subscription) },
+            { "subscription_item", typeof(SubscriptionItem) },
+            { "terminal.connection_token", typeof(Terminal.ConnectionToken) },
+            { "terminal.location", typeof(Terminal.Location) },
+            { "terminal.reader", typeof(Terminal.Reader) },
+            { "three_d_secure", typeof(ThreeDSecure) },
+            { "token", typeof(Token) },
+            { "topup", typeof(Topup) },
+            { "transfer", typeof(Transfer) },
+            { "transfer_reversal", typeof(TransferReversal) },
+            { "usage_record", typeof(UsageRecord) },
+            { "usage_record_summary", typeof(UsageRecordSummary) },
+            { "webhook_endpoint", typeof(WebhookEndpoint) },
+        };
+
+        /// <summary>
+        /// Returns the concrete type to use, given a potential type and the value of the `object`
+        /// key in a JSON payload.
+        /// </summary>
+        /// <param name="potentialType">Potential type. Can be a concrete type or an interface.</param>
+        /// <param name="objectValue">Value of the `object` key in the JSON payload.</param>
+        /// <returns>The concrete type to use, or `null`.</returns>
+        public static Type GetConcreteType(Type potentialType, string objectValue)
+        {
+            if (potentialType != null && !potentialType.GetTypeInfo().IsInterface)
+            {
+                // Potential type is already a concrete type, return it.
+                return potentialType;
+            }
+
+            Type concreteType = null;
+
+            if (!string.IsNullOrEmpty(objectValue) &&
+                ObjectsToTypes.TryGetValue(objectValue, out concreteType))
+            {
+                // Found a concrete type matching the value of the `object` key, check if it's
+                // compatible with the interface.
+                if (potentialType.GetTypeInfo().IsAssignableFrom(concreteType.GetTypeInfo()))
+                {
+                    return concreteType;
```

Comment:
how could this one really happen? Since to exist in your Dictionary above you have to have a mapping to the corresponding class

Reply:
Let's say we're trying to deserialize an `IExternalAccount`, but the API returned `"object": "source"`. The mapping would return `Source` as the concrete type, but `Source` does not implement `IExternalAccount`. We could instantiate the type and populate it, but then the deserializer would crash when trying to assign the value to the property, since the types are incompatible.